### PR TITLE
Add documentation about context builder

### PR DESCRIPTION
**Pull request by vincentchalamon**

### CODE REVIEW

File: core/serialization-groups-and-relations.md [hunk 1]
```diff
@@ -235,6 +235,101 @@ the data provider and any changes in the embedded relation will be applied to th
 
 You can create as relation embedding levels as you want.
 
+### Extending serialization context
```
Comment (dunglas) [inline core/serialization-groups-and-relations.md]: I suggest: `Change Dynamically the Serialization Context`. And please use title case for titles.

File: core/serialization-groups-and-relations.md [hunk 2]
```diff
@@ -235,6 +235,101 @@ the data provider and any changes in the embedded relation will be applied to th
...
+
+It is possible to dynamically extend serialization context. Let's imagine a resource where most fields can be managed by
```
Comment (dunglas) [inline core/serialization-groups-and-relations.md]: the serialization context?

File: core/serialization-groups-and-relations.md [hunk 3]
```diff
@@ -235,6 +235,101 @@ the data provider and any changes in the embedded relation will be applied to th
...
+     */
+    private $name;
+}
+```
+
+All entry points are the same for all users, so we should find a way to detect if authenticated user is admin, and if so
+dynamically add `admin_input` to deserialization groups.
+
+API Platform implements a `ContextBuilder`, which prepares context for serialization & deserialization. Let's [decorate
+this service](http://symfony.com/doc/current/service_container/service_decoration.html) to override `createFromRequest`
```
Comment (dunglas) [inline core/serialization-groups-and-relations.md]: the `createRequest`...?

File: core/serialization-groups-and-relations.md [hunk 4]
```diff
@@ -235,6 +235,101 @@ the data provider and any changes in the embedded relation will be applied to th
...
+    app.serializer.builder.book:
+        decorates: api_platform.serializer.context_builder
+        class: AppBundle\Serializer\BookContextBuilder
+        arguments: ['@app.serializer.builder.book.inner', '@security.authorization_checker']
+```
+
+```php
+// src/AppBundle/Serializer/BookContextBuilder.php
+
+namespace LesTilleuls\TournamentBundle\Serializer;
```
Comment (dunglas) [inline core/serialization-groups-and-relations.md]: Should be `AppBundle\Serializer`.

File: core/serialization-groups-and-relations.md [hunk 5]
```diff
@@ -235,6 +235,101 @@ the data provider and any changes in the embedded relation will be applied to th
...
+
+use ApiPlatform\Core\Serializer\SerializerContextBuilderInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+
+class BookContextBuilder implements SerializerContextBuilderInterface
```
Comment (dunglas) [inline core/serialization-groups-and-relations.md]: Can you make this class `final` (according to our Best Practices)?

**Comment by dunglas** [comment]
Great doc!
I left some minor comments. Can you also add the entry to the TOC?

Thank you very much.

File: core/serialization-groups-and-relations.md (hunk 3 above)
Comment (vincentchalamon) [inline core/serialization-groups-and-relations.md]: Method name is `createFromRequest`: https://github.com/api-platform/core/blob/2d972c0a54afdc21f68c744895c394edb8cd4d58/src/Serializer/SerializerContextBuilderInterface.php#L35

**Comment by dunglas** [comment]
👍

File: core/serialization-groups-and-relations.md (hunk 1 above)
Comment (teohhanhui) [inline core/serialization-groups-and-relations.md]: Changing the Serialization Context Dynamically

File: core/serialization-groups-and-relations.md [hunk 6]
```diff
@@ -235,6 +235,100 @@ the data provider and any changes in the embedded relation will be applied to th
 
 You can create as relation embedding levels as you want.
 
+### Change Dynamically the Serialization Context
```
Comment (teohhanhui) [inline core/serialization-groups-and-relations.md]: Changing the Serialization Context Dynamically

File: core/serialization-groups-and-relations.md [hunk 7]
```diff
@@ -235,6 +235,100 @@ the data provider and any changes in the embedded relation will be applied to th
...
+     * @Groups({"book_output", "book_input"})
+     */
+    private $name;
+}
+```
+
+All entry points are the same for all users, so we should find a way to detect if authenticated user is admin, and if so
+dynamically add `admin_input` to deserialization groups.
+
+API Platform implements a `ContextBuilder`, which prepares context for serialization & deserialization. Let's [decorate
```
Comment (teohhanhui) [inline core/serialization-groups-and-relations.md]: the context

File: core/serialization-groups-and-relations.md [hunk 8]
```diff
@@ -235,6 +235,100 @@ the data provider and any changes in the embedded relation will be applied to th
...
+        if ($this->authorizationChecker->isGranted('ROLE_ADMIN') && false === $normalization) {
+            $context['groups'][] = 'admin_input';
+        }
+
+        return $context;
+    }
+}
+```
+
+If user has `ROLE_ADMIN` permission, `admin_input` group will be dynamically added to denormalization context. Variable
```
Comment (teohhanhui) [inline core/serialization-groups-and-relations.md]: * the user
* the denormalization context
* The variable
* lets you check whether the context is for normalization

**Comment by Simperfit** [comment]
👍

**Comment by dunglas** [comment]
Thanks Vincent!

**Pull request merged**